File: pro++/sqlite_utils.cpp
//
//  sqlite数据库使用函数
//

#include "sqlite_utils.h"

sqlite3* pDB = NULL;

/*
 *  author: 无名
 *  date  : 2018-08-07
 *  note  : close sqlite connection
 */
int sqlite_close() {
	sqlite3_close(pDB);
	return 0;
}


/*
*  author: 无名
*  date  : 2018-08-07
*  note  : connect to sqlite
*/
int sqlite_conn() {
	int nRes = sqlite3_open("data.db", &pDB);
	if (nRes != SQLITE_OK)
		return -1;
	return 0;
}


/*
*  author: 无名
*  date  : 2019-12-18
*  note  : create tables if not exists
*/
bool create_tables() {
	using namespace std;
	//id,writing_pro,reading_pro,art_learning_pro,total_pro,note,str_date,date
	string str_sql = "CREATE TABLE IF NOT EXISTS  pro(id integer  primary key autoincrement,writing_pro integer,reading_pro integer,art_learning_pro integer,total_pro integer,note varchar(500),date date NOT NULL); ";
	char* cErrMsg;
	int nRes = sqlite3_exec(pDB, str_sql.c_str(), 0, 0, &cErrMsg);
	if (nRes != SQLITE_OK)
		return false;
	return true;
}


bool add_pro(CPro pro, std::wstring date) {
	using namespace std;
	//time_t t = time(0);
	//wchar_t ch_time[20];
	//wcsftime(ch_time, sizeof(ch_time)/2, L"%Y-%m-%d", localtime(&t));

	int pos;
	pos = date.find(L"/");
	while (pos != -1) {
		// str.length()求字符的长度，注意str必须是string类型
		date.replace(pos, wstring(L"/").length(), L"-");
		pos = date.find(L"/");
	}

	wstring sql = L"INSERT INTO pro(writing_pro, reading_pro, art_learning_pro, total_pro,note,date) VALUES ('";
	sql.append(pro.get_writing_pro_str()).append(L"','").append(pro.get_reading_pro_str()).append(L"','").append(pro.get_art_learning_pro_str());
	sql.append(L"','").append(pro.get_total_pro_str()).append(L"','").append(pro.get_note()).append(L"','").append(date).append(L"');");

	std::string str_sql;
	wchar_to_string(str_sql, sql.c_str());
	// sqlite使用utf-8编码，汉字需要转
	string str_utf8_sql = ascii_2_utf8(str_sql);

	char* errmsg;
	if (sqlite3_exec(pDB, str_utf8_sql.c_str(), NULL, NULL, &errmsg) != SQLITE_OK) {
		cout << "sqlite : insert data failed. error : " << errmsg << endl;
		sqlite3_free(errmsg);
		return false;
	}
	return true;
}


bool update_pro(CPro pro, std::wstring date) {
	using namespace std;

	wstring sql = L"update pro set writing_pro='";
	sql.append(pro.get_writing_pro_str()).append(L"',reading_pro='").append(pro.get_reading_pro_str()).append(L"',art_learning_pro='")
		.append(pro.get_art_learning_pro_str()).append(L"',total_pro='").append(pro.get_total_pro_str()).append(L"',note='").append(pro.get_note()).append(L"' where strftime('%Y/%m/%d',date)='");
	sql.append(date);
	sql.append(L"';");

	string str_sql;
	wchar_to_string(str_sql, sql.c_str());
	// sqlite使用utf-8编码，汉字需要转
	string str_utf8_sql = ascii_2_utf8(str_sql);

	char* errmsg;
	if (sqlite3_exec(pDB, str_utf8_sql.c_str(), NULL, NULL, &errmsg) != SQLITE_OK) {
		cout << "sqlite : insert data failed. error : " << errmsg << endl;
		sqlite3_free(errmsg);
		return false;
	}
	return true;
}


bool get_pro_sum_by_month(float* const writing_pro_sum, float* const reading_pro_sum, 
	float* const art_learning_pro_sum, float* const total_pro_sum, const int year, const int month) {
	using namespace std;
	char year_buf[15];
	char month_buf[15];
	_itoa(year,year_buf,10);
	_itoa(month,month_buf,10);
	string sql = "select sum(writing_pro),sum(reading_pro),sum(art_learning_pro),sum(total_pro) from pro where strftime('%Y-%m',date)='";
	sql.append(year_buf).append("-").append(month_buf).append("';");

	char* errmsg;
	char** pResult;
	int nRow, nCol;
	if (sqlite3_get_table(pDB, sql.c_str(), &pResult, &nRow, &nCol, &errmsg) != SQLITE_OK) {
		cout << "sqlite : select data failed. error : " << errmsg << endl;
		sqlite3_free(errmsg);
		return false;
	}
	int nIndex = nCol;
	if (pResult[nIndex] == NULL || strcmp(pResult[nIndex], "") == 0)
	{
		*writing_pro_sum = 0;
	}
	else 
	{
		*writing_pro_sum = static_cast<float>(atof(pResult[nIndex]));
	}
	if (pResult[nIndex+1] == NULL || strcmp(pResult[nIndex+1], "") == 0)
	{
		*reading_pro_sum = 0;
	}
	else
	{
		*reading_pro_sum = static_cast<float>(atof(pResult[nIndex + 1]));
	}
	if (pResult[nIndex + 2] == NULL || strcmp(pResult[nIndex + 2], "") == 0)
	{
		*art_learning_pro_sum = 0;
	}
	else 
	{
		*art_learning_pro_sum = static_cast<float>(atof(pResult[nIndex + 2]));
	}
	if (pResult[nIndex + 3] == NULL || strcmp(pResult[nIndex + 3], "") == 0)
	{
		*total_pro_sum = 0;
	}
	else
	{
		*total_pro_sum = static_cast<float>(atof(pResult[nIndex + 3]));
	}
	return true;
}


bool get_pro_sum_by_year(float* const writing_pro_sum, float* const reading_pro_sum,
	float* const art_learning_pro_sum, float* const total_pro_sum, const int year) {
	using namespace std;
	char year_buf[15];
	_itoa(year, year_buf, 10);
	string sql = "select sum(writing_pro),sum(reading_pro),sum(art_learning_pro),sum(total_pro) from pro where strftime('%Y',date)='";
	sql.append(year_buf).append("';");

	char* errmsg;
	char** pResult;
	int nRow, nCol;
	if (sqlite3_get_table(pDB, sql.c_str(), &pResult, &nRow, &nCol, &errmsg) != SQLITE_OK) {
		cout << "sqlite : select data failed. error : " << errmsg << endl;
		sqlite3_free(errmsg);
		return false;
	}
	int nIndex = nCol;
	if (pResult[nIndex] == NULL || strcmp(pResult[nIndex], "") == 0)
	{
		*writing_pro_sum = 0;
	}
	else
	{
		*writing_pro_sum = static_cast<float>(atof(pResult[nIndex]));
	}
	if (pResult[nIndex + 1] == NULL || strcmp(pResult[nIndex + 1], "") == 0)
	{
		*reading_pro_sum = 0;
	}
	else
	{
		*reading_pro_sum = static_cast<float>(atof(pResult[nIndex + 1]));
	}
	if (pResult[nIndex + 2] == NULL || strcmp(pResult[nIndex + 2], "") == 0)
	{
		*art_learning_pro_sum = 0;
	}
	else
	{
		*art_learning_pro_sum = static_cast<float>(atof(pResult[nIndex + 2]));
	}
	if (pResult[nIndex + 3] == NULL || strcmp(pResult[nIndex + 3], "") == 0)
	{
		*total_pro_sum = 0;
	}
	else
	{
		*total_pro_sum = static_cast<float>(atof(pResult[nIndex + 3]));
	}
	return true;
}


bool get_pro_sum(float* const writing_pro_sum, float* const reading_pro_sum,
	float* const art_learning_pro_sum, float* const total_pro_sum) {
	using namespace std;
	string sql = "select sum(writing_pro),sum(reading_pro),sum(art_learning_pro),sum(total_pro) from pro;";

	char* errmsg;
	char** pResult;
	int nRow, nCol;
	if (sqlite3_get_table(pDB, sql.c_str(), &pResult, &nRow, &nCol, &errmsg) != SQLITE_OK) {
		cout << "sqlite : select data failed. error : " << errmsg << endl;
		sqlite3_free(errmsg);
		return false;
	}
	int nIndex = nCol;
	if (pResult[nIndex] == NULL || strcmp(pResult[nIndex], "") == 0)
	{
		*writing_pro_sum = 0;
	}
	else
	{
		*writing_pro_sum = static_cast<float>(atof(pResult[nIndex]));
	}
	if (pResult[nIndex + 1] == NULL || strcmp(pResult[nIndex + 1], "") == 0)
	{
		*reading_pro_sum = 0;
	}
	else
	{
		*reading_pro_sum = static_cast<float>(atof(pResult[nIndex + 1]));
	}
	if (pResult[nIndex + 2] == NULL || strcmp(pResult[nIndex + 2], "") == 0)
	{
		*art_learning_pro_sum = 0;
	}
	else
	{
		*art_learning_pro_sum = static_cast<float>(atof(pResult[nIndex + 2]));
	}
	if (pResult[nIndex + 3] == NULL || strcmp(pResult[nIndex + 3], "") == 0)
	{
		*total_pro_sum = 0;
	}
	else
	{
		*total_pro_sum = static_cast<float>(atof(pResult[nIndex + 3]));
	}
	return true;
}


bool check_if_pro_of_date_exist(std::wstring date) {
	using namespace std;
	wstring sql = L"select count(*) from pro where strftime('%Y/%m/%d',date)='";
	sql.append(date);
	sql.append(L"';");
	std::string str_sql;
	wchar_to_string(str_sql, sql.c_str());
	char* errmsg;
	char** pResult;
	int nRow, nCol;
	if (sqlite3_get_table(pDB, str_sql.c_str(), &pResult, &nRow, &nCol, &errmsg) != SQLITE_OK) {
		cout << "sqlite : select data failed. error : " << errmsg << endl;
		sqlite3_free(errmsg);
		return false;
	}
	int nIndex = nCol;
	if (atoi(pResult[nIndex]) > 0)
		return true;
	return false;
}


bool get_pro(std::wstring date, CPro* pro) {
	using namespace std;
	wstring sql = L"select writing_pro,reading_pro,art_learning_pro,note from pro where strftime('%Y/%m/%d',date)='";
	sql.append(date).append(L"';");
	string str_sql;
	wchar_to_string(str_sql, sql.c_str());
	char* errmsg;
	char** pResult;
	int nRow, nCol;
	if (sqlite3_get_table(pDB, str_sql.c_str(), &pResult, &nRow, &nCol, &errmsg) != SQLITE_OK) {
		cout << "sqlite : select data failed. error : " << errmsg << endl;
		sqlite3_free(errmsg);
		return false;
	}
	if (nRow == 0 || nCol == 0)
		return false;
	int nIndex = nCol;
	float writing_pro_num, reading_pro_num, art_learning_pro_num = 0.0;
	if (pResult[nIndex] == NULL || strcmp(pResult[nIndex], "") == 0)
	{
		writing_pro_num = 0;
	}
	else
	{
		writing_pro_num = static_cast<float>(atof(pResult[nIndex]));
	}
	if (pResult[nIndex + 1] == NULL || strcmp(pResult[nIndex + 1], "") == 0)
	{
		reading_pro_num = 0;
	}
	else
	{
		reading_pro_num = static_cast<float>(atof(pResult[nIndex + 1]));
	}
	if (pResult[nIndex + 2] == NULL || strcmp(pResult[nIndex + 2], "") == 0)
	{
		art_learning_pro_num = 0;
	}
	else
	{
		art_learning_pro_num = static_cast<float>(atof(pResult[nIndex + 2]));
	}
	string str_note;
	if (pResult[nIndex + 3] == NULL || strcmp(pResult[nIndex + 3], "") == 0)
	{
		str_note = "";
	}
	else
	{
		str_note = pResult[nIndex + 3];
		str_note = utf8_2_ascii(str_note);
	}
	wstring wstr_note;
	string_to_wstring(wstr_note, str_note);
	CPro new_pro(writing_pro_num, art_learning_pro_num,reading_pro_num, wstr_note);
	//浅拷贝即可
	*pro = new_pro;
	return true;
}

File: python_cmd_version/sqlite_utils.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-
import sys

import sqlite3
from pro_entity import pro

reload(sys)
sys.setdefaultencoding( "utf-8" )

class sqlite_utils:
    
    conn = None
    
    def conn_sqlite(self): 
        self.conn = sqlite3.connect('data.db')
        #print "Opend db successfully"
        
    def close_sqlite(self):
        self.conn.close()   
        
    def create_tables(self):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        c.execute('''CREATE TABLE IF NOT EXISTS  pro(id integer  primary key autoincrement,writing_pro integer,reading_pro integer,art_learning_pro integer,computer_science_tech_pro integer,total_pro integer,note varchar(500),date date NOT NULL);''')  
        #print "Table created successfully"
        self.conn.close()   
    
    def add_pro(self,pro,date):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str="INSERT INTO pro(writing_pro, reading_pro, art_learning_pro, computer_science_tech_pro, total_pro, note, date) VALUES ('"
        sql_str = sql_str + pro.get_writing_pro_str() + "','" + pro.get_reading_pro_str() + "','" + pro.get_art_learning_pro_str() + "','" + pro.get_computer_learning_pro_str() + "','" + pro.get_total_pro_str() + "','" + pro.get_note() + "','" + date+ "');"
        self.conn.execute(sql_str)
        self.conn.commit()
        #print("sql is"+sql_str)
        self.conn.close()   

    def sel_pro_by_date(self,str_date):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str="select writing_pro,reading_pro,art_learning_pro,computer_science_tech_pro,note from pro where date='"
        sql_str = sql_str + str_date + "';"
        #print('sql:'+sql_str)
        cursor = self.conn.execute(sql_str)
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return None
        row = cursor_list[0]
        
        row_list = list(row)
        if row[0] == None or row[0] == '':
            row_list.append('0')
        else:
            row_list.append(row[0])
        if row[1] == None or row[1] == '':
            row_list.append('0')
        else:
            row_list.append(row[1])
        if row[2] == None or row[2] == '':
            row_list.append('0')
        else:
            row_list.append(row[2])
        if row[3] == None or row[3] == '':
            row_list.append('0')
        else:
            row_list.append(row[3])
        if row[4] == None or row[4] == '':
            row_list.append('')
        else:
            row_list.append(row[4])
            
        #print(row)
        pro_sel = pro(float(row_list[0]),float(row_list[2]),float(row_list[1]),float(row_list[3]),row_list[4])
        self.conn.close()   
        return pro_sel
        
    def update_pro(self,pro,date):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str = "update pro set writing_pro='"
        sql_str = sql_str+pro.get_writing_pro_str() +"',reading_pro='"+pro.get_reading_pro_str()+"',art_learning_pro='"+pro.get_art_learning_pro_str()+"',computer_science_tech_pro='"+pro.get_computer_learning_pro_str()+"',total_pro='"+pro.get_total_pro_str()+"',note='"+pro.get_note()+"' where date='"+date+"';"
        self.conn.execute(sql_str)
        self.conn.commit()
        #print("sql is"+sql_str)
        self.conn.close() 
        
    def check_if_pro_of_date_exist(self,date):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str="select count(*) from pro where date='"+date+"';"
        cursor = self.conn.execute(sql_str)
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return None
        row = cursor_list[0]
        #print(row)
        self.conn.close()  
        if int(row[0]) > 0:
            return True
        return False    
        
    def get_pro_sum_by_month(self,year,month):
        ret_sum_nums = []
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str = "select sum(writing_pro),sum(reading_pro),sum(art_learning_pro),sum(computer_science_tech_pro),sum(total_pro) from pro where strftime('%Y-%m',date)='";
        if month<10:
            sql_str = sql_str + str(year) + "-0" + str(month) + "';"
        else:
            sql_str = sql_str + str(year) + "-" + str(month) + "';"
        cursor = self.conn.execute(sql_str)
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return None
        row = cursor_list[0]
        #print(row)
        if row[0] == None or row[0] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[0]))
        if row[1] == None or row[1] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[1]))
        if row[2] == None or row[2] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[2]))
        if row[3] == None or row[3] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[3]))
        if row[4] == None or row[4] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[4]))
        #print(ret_sum_nums)
        self.conn.close()     
        return ret_sum_nums        

    def count_pro_by_month(self,year,month):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str="select count(*) from pro where strftime('%Y-%m',date)='"
        if month < 10:
            sql_str = sql_str + str(year) +  '-' +  '0' + str(month)+ "';"
        else:
            sql_str = sql_str + str(year) +  '-'  + str(month) +"';"
        cursor = self.conn.execute(sql_str)
        if cursor == None:
            return 0
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return 0
        row = cursor_list[0]
        #print(row)
        self.conn.close()  
        return float(row[0])
        
    def get_pro_sum_by_year(self,year):
        ret_sum_nums = []
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str = "select sum(writing_pro),sum(reading_pro),sum(art_learning_pro),sum(computer_science_tech_pro),sum(total_pro) from pro where strftime('%Y',date)='"  + str(year) + "';";
        #print('sql:'+sql_str)
        cursor = self.conn.execute(sql_str)
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return None
        row = cursor_list[0]
        #print(row)
        if row[0] == None or row[0] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[0]))
        if row[1] == None or row[1] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[1]))
        if row[2] == None or row[2] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[2]))
        if row[3] == None or row[3] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[3]))
        if row[4] == None or row[4] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[4]))
        #print(ret_sum_nums)
        self.conn.close()     
        return ret_sum_nums        

    def count_pro_by_year(self,year):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str="select count(*) from pro where strftime('%Y',date)='" + str(year) + "';"
        cursor = self.conn.execute(sql_str)
        if cursor == None:
            return 0
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return 0
        row = cursor_list[0]
        #print(row)
        self.conn.close()  
        return float(row[0])
        
    def get_pro_sum(self):
        ret_sum_nums = []
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str = "select sum(writing_pro),sum(reading_pro),sum(art_learning_pro),sum(computer_science_tech_pro),sum(total_pro) from pro;";
        #print('sql:'+sql_str)
        cursor = self.conn.execute(sql_str)
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return None
        row = cursor_list[0]
        #print(row)
        if row[0] == None or row[0] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[0]))
        if row[1] == None or row[1] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[1]))
        if row[2] == None or row[2] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[2]))
        if row[3] == None or row[3] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[3]))
        if row[4] == None or row[4] == '':
            ret_sum_nums.append(0.00)
        else:
            ret_sum_nums.append(float(row[4]))
        #print(ret_sum_nums)
        self.conn.close()     
        return ret_sum_nums        

    def count_pro(self):
        self.conn = sqlite3.connect('data.db')
        c = self.conn.cursor()
        sql_str="select count(*) from pro;"
        cursor = self.conn.execute(sql_str)
        if cursor == None:
            return 0
        cursor_list = list(cursor)
        if len(cursor_list)<1:
            return 0
        row = cursor_list[0]
        #print(row)
        self.conn.close()  
        return float(row[0])
        